File: bench/cpp/push.cpp
// Growth while holding element pointers: see bench/goose/push.goose. A
// vector's elements move on every reallocation, so the marks have to be
// indices unless the container is chosen for pointer stability.
//   VARIANT 0: vector, no reserve, marks as indices.
//   VARIANT 1: vector with an exact reserve, marks as indices.
//   VARIANT 2: deque (pointer-stable), marks as real pointers.
#include "bench.h"
#include <vector>
#include <deque>

struct Item { int32_t key, val; };

int main() {
    uint64_t r = 12345;
    long long total = 0, mt = 0, nitems = 0, nmarks = 0;
#if VARIANT == 2
    std::deque<Item> items;
    std::vector<Item *> marks;
    for (long long i = 0; i < BENCH_N; i++) {
        r = xs_next(r);
        items.push_back(Item { (int32_t)xs_mod(r, 1000000), (int32_t)(i % 97) });
        if (i % 64 == 0) marks.push_back(&items.back());
    }
    for (const auto &it : items) total += it.key;
    for (auto *m : marks) mt += m->val;
#else
    std::vector<Item> items;
    std::vector<size_t> marks;
  #if VARIANT == 1
    items.reserve(BENCH_N);
  #endif
    for (long long i = 0; i < BENCH_N; i++) {
        r = xs_next(r);
        items.push_back(Item { (int32_t)xs_mod(r, 1000000), (int32_t)(i % 97) });
        if (i % 64 == 0) marks.push_back(items.size() - 1);
    }
    for (const auto &it : items) total += it.key;
    for (size_t m : marks) mt += items[m].val;
#endif
    nitems = (long long)items.size();
    nmarks = (long long)marks.size();
    emit(nitems);
    emit(nmarks);
    emit(total);
    emit(mt);
    return 0;
}

File: bench/rust/push.rs
// Growth while holding element pointers: see bench/goose/push.goose. Safe Rust
// has no answer to this at all -- a `&Item` into `items` borrows the vector for
// as long as it is held, so the next `push` will not compile, and no std
// container gives both pointer stability and O(1) append. The marks are
// therefore indices, which is not a performance choice but the only shape the
// language admits. The capacity is reserved because N is a known constant here
// and a Rust programmer would write that; it keeps the benchmark about the
// pointers-versus-indices question rather than about reallocation.
mod bench;
use bench::*;

const N: i64 = 1000000;                  // BENCH_N

struct Item { key: i32, val: i32 }

fn main() {
    let mut items: Vec<Item> = Vec::with_capacity(N as usize);
    let mut marks: Vec<usize> = Vec::new();
    let mut r: u64 = 12345;
    for i in 0..N {
        r = xs_next(r);
        items.push(Item { key: xs_mod(r, 1000000) as i32, val: (i % 97) as i32 });
        if i % 64 == 0 { marks.push(items.len() - 1); }
    }
    let mut total: i64 = 0;
    for it in &items { total += it.key as i64; }
    let mut mt: i64 = 0;
    for &m in &marks { mt += items[m].val as i64; }
    emit(items.len() as i64);
    emit(marks.len() as i64);
    emit(total);
    emit(mt);
}
